### PR TITLE
protect read-write-extend operations with a lock

### PR DESCRIPTION
**Pull request by lum1n0us**

### CODE REVIEW

File: core/shared/utils/bh_vector.c [hunk 1]
```diff
@@ -65,6 +80,15 @@ bh_vector_init(Vector *vector, size_t init_length, size_t size_elem)
     vector->size_elem = size_elem;
     vector->max_elems = init_length;
     vector->num_elems = 0;
+
+    if (use_lock) {
+        vector->lock = wasm_runtime_malloc(sizeof(korp_mutex));
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:85]: Should check the return value

File: core/shared/utils/bh_vector.c [hunk 2]
```diff
@@ -65,6 +80,15 @@ bh_vector_init(Vector *vector, size_t init_length, size_t size_elem)
...
+        os_mutex_init(vector->lock);
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:86]: Should check the return value

File: core/shared/utils/bh_vector.c [hunk 3]
```diff
@@ -5,6 +5,18 @@
 
 #include "bh_vector.h"
 
+#define LOCK(v)               \
+    {                         \
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:9]: Had better wrap the code with `do { .. } while (0)`

File: core/shared/utils/bh_vector.c [hunk 4]
```diff
@@ -5,6 +5,18 @@
 
 #include "bh_vector.h"
 
+#define LOCK(v)               \
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:8]: code guideline check for this file failed

File: core/shared/utils/bh_vector.c [hunk 5]
```diff
@@ -65,6 +81,22 @@ bh_vector_init(Vector *vector, size_t init_length, size_t size_elem)
...
+    if (use_lock) {
+        if (!(vector->lock = wasm_runtime_malloc(sizeof(korp_mutex)))) {
+            LOG_ERROR("Init vector failed: alloc locker failed.\n");
+            bh_vector_destroy(vector);
+            return false;
+        }
+
+        if (BHT_ERROR == os_mutex_init(vector->lock)) {
+            LOG_ERROR("Init vector failed: init locker failed.\n");
+            bh_vector_destroy(vector);
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:95]: Should `wasm_runtime_free(vector->lock) and set vector->lock to NULL` before destroy vector, or in bh_vector_destroy, L249, os_mutex_destroy will destroy an uninitialized lock.

File: core/shared/utils/bh_vector.c [hunk 6]
```diff
@@ -65,6 +81,22 @@ bh_vector_init(Vector *vector, size_t init_length, size_t size_elem)
...
+    vector->lock = NULL;
+
+    if (use_lock) {
+        if (!(vector->lock = wasm_runtime_malloc(sizeof(korp_mutex)))) {
+            LOG_ERROR("Init vector failed: alloc locker failed.\n");
+            bh_vector_destroy(vector);
+            return false;
+        }
+
+        if (BHT_ERROR == os_mutex_init(vector->lock)) {
```
Comment (wenyongh) [inline core/shared/utils/bh_vector.c:93]: Should use `0 != os_mutex_init` or `BHT_OK != os_mutex_init`, as os_mutex_init is implemented by various platforms, the return value may be not BHT_ERROR when we failed to init it.

File: core/shared/utils/bh_vector.c (hunk 5 above)
Comment (lum1n0us) [reply core/shared/utils/bh_vector.c:95]: I see. nice catch 👍

File: core/shared/utils/bh_vector.c (hunk 6 above)
Comment (lum1n0us) [reply core/shared/utils/bh_vector.c:93]: got it

File: doc/wasm_c_api.md [hunk 7]
```diff
@@ -3,24 +3,42 @@
...
 
 And here are [examples](https://github.com/WebAssembly/wasm-c-api/tree/master/example) which
 are helpful.
 
+## FYI
+
+- The thread model of _wasm_c_api_ is
+
+  - An `wasm_engine_t` instance may only be created once per process
+  - Every `wasm_store_t` and its objects only be accessed in a single thread
```
Comment (wenyongh) [inline doc/wasm_c_api.md:17]: `may` only be accessed

File: doc/wasm_c_api.md [hunk 8]
```diff
@@ -3,24 +3,42 @@
...
+
+- `wasm_engine_new`, `wasm_engine_new_with_config`, `wasm_engine_new_with_args`,
+  `wasm_engine_delete`should be called in a thread-safe environment. Such
+  behaviors are not recommend , and please make sure an appropriate calling
```
Comment (wenyongh) [inline doc/wasm_c_api.md:21]: `are not recommend ,` to `are not recommended,`

File: doc/wasm_c_api.md [hunk 9]
```diff
@@ -3,24 +3,42 @@
...
+  sequence if it has to be.
+
+  - to call `wasm_engine_new` and `wasm_engine_delete` in different threads
```
Comment (wenyongh) [inline doc/wasm_c_api.md:24]: `to call` means do it in the future, had better change it to `call`?

File: doc/wasm_c_api.md [hunk 10]
```diff
@@ -3,24 +3,42 @@
...
+  - to call `wasm_engine_new` or `wasm_engine_delete` multiple time in
```
Comment (wenyongh) [inline doc/wasm_c_api.md:25]: multiple times

**Pull request merged**